### PR TITLE
[Microsoft.DeviceRegistry] Added spec for Microsoft.DeviceRegistry RP version 2024-09-01-preview

### PR DESCRIPTION
**Pull request by davidemontanari**
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

spec_pr_review_workflow_diagram

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Convert existing [OpenAPI spec to TypeSpec spec](https://aka.ms/typespec/conversion) (do not combine this with implementing changes for a new API version).
  - [x] Other, please clarify:
    - Added new resource types.
    - Updated existing resource types.
    - Fixed examples reference links.
    - Regenerated swaggers using latest TypeSpec compiler.
    - Promotes the latest Private Preview API (`2024-07-01-preview`) to Public Preview:
      - PR for `2024-07-01-preview` API: https://github.com/Azure/azure-rest-api-specs-pr/pull/17529
      - Following PR for patching `2024-07-01-preview` API: https://github.com/Azure/azure-rest-api-specs-pr/pull/18510
      - TypeSpec output changed in `0.59.x` compared to previous versions:
        - Resource update output changed format
        - PR taken as reference to define the new update payload: https://github.com/Azure/azure-rest-api-specs-pr/pull/19220
        - Commit to fix the resource update payload and address the breaking changes in this current PR: https://github.com/Azure/azure-rest-api-specs/pull/30222/commits/9ea88b2f9edc31f67c29e82622d42f8b2eac36aa
      - `LintDiff` errors were present also in the previous API version (`2023-11-01-preview`) and approved in `2024-07-01-preview` Private Preview API version (link above).

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories)
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

### CODE REVIEW

**Review by marcodalessandro — APPROVED** [review]

**Comment by davidemontanari** [comment]
For the ARM reviewer:

TypeSpec output changed in `0.59.x` compared to previous versions:
- Resource update output changed format
- PR taken as reference to define the new update payload: https://github.com/Azure/azure-rest-api-specs-pr/pull/19220
- Commit to fix the resource update payload and address the breaking changes in this current PR: https://github.com/Azure/azure-rest-api-specs/pull/30222/commits/9ea88b2f9edc31f67c29e82622d42f8b2eac36aa
- `LintDiff` errors were present also in the previous API version (`2023-11-01-preview`) and approved in `2024-07-01-preview` Private Preview API version (link above).

**Comment by gary-x-li** [comment]
Can you take a look at the LintDiff errors? Some of them are not in the original PRs.
https://github.com/Azure/azure-rest-api-specs/pull/30222/checks?check_run_id=29678670766

**Comment by davidemontanari** [comment]
> Can you take a look at the LintDiff errors? Some of them are not in the original PRs. https://github.com/Azure/azure-rest-api-specs/pull/30222/checks?check_run_id=29678670766

Thanks @gary-x-li for taking a look!

Here's an analysis on the LintDiff errors:

- [PatchBodyParametersSchema](https://github.com/Azure/azure-openapi-validator/blob/main/docs/patch-body-parameters-schema.md) - Enums with default values keep the default values during PATCH as well. Looks like that TypeSpec doesn't strip out default values when the properties are passed to the update function (while required properties are correctly removed). This is something we already have today for API version `2023-11-01-preview` (Public Preview) and `2024-07-01-preview` (Private Preview). Example [here](https://github.com/Azure/azure-rest-api-specs-pr/blob/fbfbb62a816c2f26cfca7db2a6810314f1e32440/specification/deviceregistry/resource-manager/Microsoft.DeviceRegistry/preview/2024-07-01-preview/deviceregistry.json#L3471).
- [PatchPropertiesCorrespondToPutProperties](https://github.com/Azure/azure-openapi-validator/blob/main/docs/patch-properties-correspond-to-put-properties.md) - It complains that the `tags` property in PATCH body is not present in the corresponding PUT body. As far as I can see, `tags` is explicitly defined for PATCH while for PUT it is inherited from the base `TrackedResource` type. No changes have been applied on how the PATCH is defined in the tsp source code. Is there anything missing here?
- [OperationIdNounVerb](https://github.com/Azure/azure-openapi-validator/blob/main/docs/operation-id-noun-verb.md) - This is about the word `Schemas` that cannot be part of the resource operation id. We have a resource type called `Schemas` and we've been asked to suppress this kind of error when reviewing `2024-07-01-preview` API.
- [TagsAreNotAllowedForProxyResources](https://github.com/Azure/azure-openapi-validator/blob/main/docs/tags-are-not-allowed-for-proxy-resources.md) - Checking the links, 3 out of 4 point to the PATCH payload for Tracked Resources, so not sure how to address these. Instead, one alert is related to a `tags` property inside `Schema` resource property (Proxy Resource) that has been approved for `2024-07-01-preview` API version.

File: specification/deviceregistry/DeviceRegistry.Management/schemaRegistries.tsp [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+  schemaType: SchemaType;
+
+  @doc("Provisioning state of the resource.")
+  @visibility("read")
+  provisioningState?: ProvisioningState;
+
+  #suppress "@azure-tools/typespec-azure-resource-manager/arm-no-record" ""
+  @doc("Schema tags.")
+  // Allowed scenario as per user-defined tags: https://github.com/Azure/azure-openapi-validator/blob/main/docs/avoid-additional-properties.md
+  tags?: Record<string>;
```
Comment (gary-x-li) [inline specification/deviceregistry/DeviceRegistry.Management/schemaRegistries.tsp:101]: Proxy resources don't support tags

Comment (gary-x-li) [reply specification/deviceregistry/DeviceRegistry.Management/schemaRegistries.tsp:101]: @davidemontanari I believe this is related to the `TagsAreNotAllowedForProxyResources` violations. You need to make them tracked resources if you need tags support.

Comment (davidemontanari) [reply specification/deviceregistry/DeviceRegistry.Management/schemaRegistries.tsp:101]: @gary-x-li, `tags` here is defined inside the resource `properties` (accessed through `properties.tags`) and not as top level property. Tracked resources define `tags` as resource top level property (outside `properties`). Is that considered the same?

Comment (gary-x-li) [reply specification/deviceregistry/DeviceRegistry.Management/schemaRegistries.tsp:101]: Thanks for the explanation. It's different from the top level `tags` so it's fine here.

**Pull request merged**